modules/2_mnist/src/mnist.cpp

```diff
@@ -14,57 +15,88 @@ inline int readInt(std::ifstream& ifs) {
 
 void loadImages(const std::string& filepath,
                 std::vector<Mat>& images) {
-    std::ifstream ifs(filepath.c_str(), std::ios::binary);
-    CV_CheckEQ(ifs.is_open(), true, filepath.c_str());
+    MnistImageReader imgDispatcher(filepath);
 
-    int magicNum = readInt(ifs);
-    CV_CheckEQ(magicNum, 2051, "");
+    printf("Images are loaded, %d\n", imgDispatcher.count());
 
-    int numImages = readInt(ifs);
-
-    // TODO: follow "FILE FORMATS FOR THE MNIST DATABASE" specification
-    // at http://yann.lecun.com/exdb/mnist/
+    images = imgDispatcher.getAllImages();
 }
 
 void loadLabels(const std::string& filepath,
                 std::vector<int>& labels) {
-    std::ifstream ifs(filepath.c_str(), std::ios::binary);
-    CV_CheckEQ(ifs.is_open(), true, filepath.c_str());
-
-    int magicNum = readInt(ifs);
-    CV_CheckEQ(magicNum, 2049, "");
+    MnistLabelReader lblDispatcher(filepath);
 
-    int numLabels = readInt(ifs);
+    printf("Labels are loaded, %d\n", lblDispatcher.count());
 
-    // TODO: follow "FILE FORMATS FOR THE MNIST DATABASE" specification
-    // at http://yann.lecun.com/exdb/mnist/
+    labels = lblDispatcher.getAllLabels();
 }
 
 void prepareSamples(const std::vector<cv::Mat>& images, cv::Mat& samples) {
-    CV_Error(Error::StsNotImplemented, "prepareSamples");
+    // assume each image has the same constraints
+    cv::Size imgConstraints = images[0].size();
+    samples = cv::Mat(images.size(), imgConstraints.width * imgConstraints.height, CV_8U);
+
+    for (int i = 0; i < images.size(); i++) {
+        images[i].reshape(1, 1).copyTo(samples.row(i));
+    }
+
+    samples.convertTo(samples, CV_32F);
 }
 
 Ptr<ml::KNearest> train(const std::vector<cv::Mat>& images,
                         const std::vector<int>& labels) {
-    CV_Error(Error::StsNotImplemented, "train");
+    Ptr<ml::KNearest> knn(ml::KNearest::create());
+    knn->setDefaultK(3);
```

Comment:
Just insteresting what is default value before it?
```
knn->getDefaultK()
```

Reply:
According to https://github.com/opencv/opencv/blob/master/modules/ml/src/knearest.cpp:69 (fef7fc3), the default value is 10. But I can see that if setting K more than 3, the execution time increases super highly, so I've decided to set it to 3.

Reply:
@dkurt , could you please take a look at mnist.cpp:77 ? 
I can't manage what do I do wrong (test **ocr** doesn't pass)

Reply:
@impulsgraw , please try to keep default KNearest classifier (the test is not perfect and it's expected to have the model without changes).